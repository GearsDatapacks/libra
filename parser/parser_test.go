package parser_test

import (
	"testing"

	"github.com/gearsdatapacks/libra/lexer"
	"github.com/gearsdatapacks/libra/lexer/token"
	"github.com/gearsdatapacks/libra/parser"
	"github.com/gearsdatapacks/libra/parser/ast"
	utils "github.com/gearsdatapacks/libra/test_utils"
)

func getProgram(t *testing.T, input string) *ast.Program {
	l := lexer.New(input, "test.lb")
	tokens := l.Tokenise()

	p := parser.New(tokens, l.Diagnostics)
	program := p.Parse()
	utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 0)

	return program
}

func getExpr[T ast.Expression](t *testing.T, program *ast.Program) T {
	utils.AssertEq(t, len(program.Statements), 1)
	stmt, ok := program.Statements[0].(*ast.ExpressionStatement)
	utils.Assert(t, ok)

	expr, ok := stmt.Expression.(T)
	utils.Assert(t, ok)
  return expr
}

func TestIdentifierExpression(t *testing.T) {
	input := "hello_World123"
  program := getProgram(t, input)

  ident := getExpr[*ast.Identifier](t, program)

	utils.AssertEq(t, ident.Name, input)
	utils.AssertEq(t, ident.Token,
		token.New(token.IDENTIFIER, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestIntegerExpression(t *testing.T) {
	input := "156_098"
	val := 156098

  program := getProgram(t, input)

  integer := getExpr[*ast.IntegerLiteral](t, program)

	utils.AssertEq(t, integer.Value, int64(val))
	utils.AssertEq(t, integer.Token,
		token.New(token.INTEGER, "156098",
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestFloatExpression(t *testing.T) {
	input := "3.1415"
	val := 3.1415

  program := getProgram(t, input)

  float := getExpr[*ast.FloatLiteral](t, program)

	utils.AssertEq(t, float.Value, val)
	utils.AssertEq(t, float.Token,
		token.New(token.FLOAT, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestBooleanExpression(t *testing.T) {
	input := "true"

  program := getProgram(t, input)

  boolean := getExpr[*ast.BooleanLiteral](t, program)

	utils.AssertEq(t, boolean.Value, true)
	utils.AssertEq(t, boolean.Token,
		token.New(token.IDENTIFIER, input,
			token.NewSpan(0, 0, len(input)),
		),
	)
}

func TestErrorExpression(t *testing.T) {
  input := ")"

  l := lexer.New(input, "test.lb")
  tokens := l.Tokenise()
  p := parser.New(tokens, l.Diagnostics)
  program := p.Parse()

  utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 1)
  diag := p.Diagnostics.Diagnostics[0]
  utils.AssertEq(t, diag.Message, "Expected expression, got `)`")
  utils.AssertEq(t, diag.Span, token.NewSpan(0, 0, 1))

  getExpr[*ast.ErrorExpression](t, program)
}

func TestMissingNewlineError(t *testing.T) {
  input := "1 2"

  l := lexer.New(input, "test.lb")
  tokens := l.Tokenise()
  p := parser.New(tokens, l.Diagnostics)
  p.Parse()

  utils.AssertEq(t, len(p.Diagnostics.Diagnostics), 1)
  diag := p.Diagnostics.Diagnostics[0]
  utils.AssertEq(t, diag.Message, "Expected newline after statement, got integer")
  utils.AssertEq(t, diag.Span, token.NewSpan(0, 2, 3))
}

