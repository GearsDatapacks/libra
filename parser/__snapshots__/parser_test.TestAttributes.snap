
[`@tag Error;struct MyError { string }` - 1]
STRUCT_DECL MyError (11:17)
├─STRUCT_FIELD
│ └─TYPE_OR_IDENT string
└─tag
  └─IDENT Error (5:10)
---

[`@impl LeInterface;fn (string) to_string(): string { this }` - 1]
FUNC_DECL to_string impl LeInterface (18:20)
├─METHOD_OF
│ └─IDENT string (22:28)
├─IDENT string (43:49)
└─BLOCK (50:51)
  └─IDENT this (52:56)
---

[`@untagged;union IntOrPtr { int: i32, ptr: *i32 }` - 1]
UNION_DECL IntOrPtr untagged (10:15)
├─UNION_MEMBER int
│ └─IDENT i32 (32:35)
└─UNION_MEMBER ptr
  └─PTR_TYPE (42:43)
    └─IDENT i32 (43:46)
---

[`@todo Implement it;fn unimplemented(param: i32) {}` - 1]
FUNC_DECL unimplemented (19:21)
├─PARAM
│ └─TYPE_OR_IDENT param
│   └─IDENT i32 (43:46)
├─BLOCK (48:49)
└─ATTRIBUTES
  └─todo = "Implement it"
---

[`@doc Does cool stuff;fn do_cool_stuff() {}` - 1]
FUNC_DECL do_cool_stuff (21:23)
├─BLOCK (40:41)
└─ATTRIBUTES
  └─doc = "Does cool stuff"
---

[`@deprecated Use `do_other_thing` instead;fn do_thing() {}` - 1]
FUNC_DECL do_thing (41:43)
├─BLOCK (55:56)
└─ATTRIBUTES
  └─deprecated = "Use `do_other_thing` instead"
---

[`@doc Has three fields;@todo Add a third field;@tag Incomplete;struct ThreeFields {i32, f32}` - 1]
STRUCT_DECL ThreeFields (62:68)
├─STRUCT_FIELD
│ └─TYPE_OR_IDENT i32
├─STRUCT_FIELD
│ └─TYPE_OR_IDENT f32
├─ATTRIBUTES
│ ├─todo = "Add a third field"
│ └─doc = "Has three fields"
└─tag
  └─IDENT Incomplete (51:61)
---

[`@deprecated;fn deprecated() {}` - 1]
FUNC_DECL deprecated (12:14)
├─BLOCK (28:29)
└─ATTRIBUTES
  └─deprecated = ""
---

[`@doc;Does some stuff.;Lorem ipsum.;I can put @end here, but if I put it on its own line,;it will close this block.;@end;fn do_things() {}` - 1]
FUNC_DECL do_things (120:122)
├─BLOCK (135:136)
└─ATTRIBUTES
  └─doc = "Does some stuff.\nLorem ipsum.\nI can put @end here, but if I put it on its own line,\nit will close this block."
---

[`@extern;fn external()` - 1]
FUNC_DECL external extern external (8:10)
---
